src/sctools/count.py

```diff
@@ -40,20 +40,101 @@ def __init__(self, matrix: sp.csr_matrix, row_index: np.ndarray, col_index: np.n
+
+    @property
+    def col_index(self):
+        return self._col_index
+
+    @staticmethod
+    def _get_alignments_grouped_by_query_name_generator(bam_file: str,
+                                                        cell_barcode_tag: str,
+                                                        molecule_barcode_tag: str,
+                                                        open_mode: str = 'rb') -> \
+            Generator[Tuple[str, Optional[str], Optional[str], List[pysam.AlignedSegment]], None, None]:
+        """Iterates through a query_name-sorted BAM file, groups all alignments with the same query name
+
+        Parameters
+        ----------
+        bam_file : str
+            input bam file marked by cell barcode, molecule barcode, and gene ID tags sorted in that
+            order
+        cell_barcode_tag : str
+            Tag that specifies the cell barcode for each read.
+        molecule_barcode_tag : str
+            Tag that specifies the molecule barcode for each read.
+
+        Returns
+        -------
+            a generator for tuples (query_name, cell_barcode, molecule_barcode, alignments)
+        """
+        with pysam.AlignmentFile(bam_file, mode=open_mode) as bam_records:
+            for alignment in itertools.groupby(bam_records, key=lambda record: record.query_name):
+                query_name: str = alignment[0]
+                grouper = alignment[1]
+                alignments: List[pysam.AlignedSegment] = []
+                try:
+                    while True:
+                        alignment = grouper.__next__()
+                        alignments.append(alignment)
+                except StopIteration:
+                    pass
+
+                cell_barcode: Optional[str] = None
+                try:
+                    cell_barcode = alignments[0].get_tag(cell_barcode_tag)
+                except KeyError:
+                    pass
+
+                molecule_barcode: Optional[str] = None
+                try:
+                    molecule_barcode = alignments[0].get_tag(molecule_barcode_tag)
+                except KeyError:
+                    pass
+
+                yield query_name, cell_barcode, molecule_barcode, alignments
+
+    # todo add support for generating a matrix of invalid barcodes
+    # todo add support for splitting spliced and unspliced reads
+    # todo add support for generating a map of cell barcodes
+    # todo add the option for stringent checks on the input (e.g. BAM sort order)
+    # todo once the stringent checks are in place, safely move on to the hashset-free implementation
     @classmethod
-    def from_bam(
+    def from_sorted_tagged_bam(
             cls,
             bam_file: str,
             annotation_file: str,
             cell_barcode_tag: str='CB',
             molecule_barcode_tag: str='UB',
-            gene_id_tag: str='GE',
-            open_mode: str='rb',
-            ):
+            gene_name_tag: str='GE',
```

Comment:
One question I forgot to ask: are these HGNC gene names that are going in the tags, and not the numerical IDs? If so then this is a good change. 

If they are IDs, I'd prefer to keep the name as ID.

Reply:
Unfortunately, they actually gene names and not IDs (we reply on `dropseq-tools` for tagging genes; it uses gene names by default, there is no way to force it to use gene IDs).